nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/storage/ADLSCredentialsControllerService.java

```diff
@@ -103,39 +106,50 @@
+        final boolean servicePrincipalSet = servicePrincipalTenantIdSet || servicePrincipalClientIdSet || servicePrincipalClientSecretSet;
 
-        boolean servicePrincipalSet = servicePrincipalTenantIdSet || servicePrincipalClientIdSet || servicePrincipalClientSecretSet;
+        final String managedIdentityClientId = validationContext.getProperty(MANAGED_IDENTITY_CLIENT_ID).getValue();
 
         if (!onlyOneSet(accountKeySet, sasTokenSet, useManagedIdentitySet, servicePrincipalSet)) {
             results.add(new ValidationResult.Builder().subject(this.getClass().getSimpleName())
                 .valid(false)
                 .explanation("one and only one authentication method of [Account Key, SAS Token, Managed Identity, Service Principal] should be used")
                 .build());
-        } else if (servicePrincipalSet) {
-            String template = "'%s' must be set when Service Principal authentication is being configured";
-            if (!servicePrincipalTenantIdSet) {
-                results.add(new ValidationResult.Builder().subject(this.getClass().getSimpleName())
-                        .valid(false)
-                        .explanation(String.format(template, SERVICE_PRINCIPAL_TENANT_ID.getDisplayName()))
-                        .build());
-            }
-            if (!servicePrincipalClientIdSet) {
-                results.add(new ValidationResult.Builder().subject(this.getClass().getSimpleName())
-                        .valid(false)
-                        .explanation(String.format(template, SERVICE_PRINCIPAL_CLIENT_ID.getDisplayName()))
-                        .build());
+        } else {
+            if (servicePrincipalSet) {
+                final String template = "'%s' must be set when Service Principal authentication is being configured";
+                if (!servicePrincipalTenantIdSet) {
+                    results.add(new ValidationResult.Builder().subject(this.getClass().getSimpleName())
+                            .valid(false)
+                            .explanation(String.format(template, SERVICE_PRINCIPAL_TENANT_ID.getDisplayName()))
+                            .build());
+                }
+                if (!servicePrincipalClientIdSet) {
+                    results.add(new ValidationResult.Builder().subject(this.getClass().getSimpleName())
+                            .valid(false)
+                            .explanation(String.format(template, SERVICE_PRINCIPAL_CLIENT_ID.getDisplayName()))
+                            .build());
+                }
+                if (!servicePrincipalClientSecretSet) {
+                    results.add(new ValidationResult.Builder().subject(this.getClass().getSimpleName())
+                            .valid(false)
+                            .explanation(String.format(template, SERVICE_PRINCIPAL_CLIENT_SECRET.getDisplayName()))
+                            .build());
+                }
```

Comment:
These are almost exactly the same so it might be more succinct to do something like below. Definitely not something I'd block on.

```suggestion
                String propertyDisplayName;
                if (!servicePrincipalTenantIdSet) {
                    propertyDisplayName = SERVICE_PRINCIPAL_TENANT_ID.getDisplayName();
                }
                if (!servicePrincipalClientIdSet) {
                    propertyDisplayName = SERVICE_PRINCIPAL_CLIENT_ID.getDisplayName();
                }
                if (!servicePrincipalClientSecretSet) {
                    propertyDisplayName = SERVICE_PRINCIPAL_CLIENT_SECRET.getDisplayName();
                }
                results.add(new ValidationResult.Builder().subject(this.getClass().getSimpleName())
                        .valid(false)
                        .explanation(String.format(template, propertyDisplayName))
                        .build());                
```

Reply:
It is not a new code but only the indentation changed here.
A minor drawback of the suggested change it that it can report only one missing property at a time (`propertyDisplayName` is overridden with the last one).
For these reasons, I would rather leave it as it is now.

Reply:
Oops. Yeah, forgot about the possibility of multiple validation errors.